src/paragraphcommand.js

```diff
@@ -69,10 +74,27 @@ export default class ParagraphCommand extends Command {
 	 * @private
 	 */
 	_updateValue() {
-		const block = this.editor.document.selection.getSelectedBlocks().next().value;
+		const block = first( this.editor.document.selection.getSelectedBlocks() );
+
+		this.value = !!block && block.is( 'paragraph' );
+	}
 
-		if ( block ) {
-			this.value = block.is( 'paragraph' );
+	/**
+	 * @inheritDoc
+	 */
+	_checkEnabled() {
+		const block = first( this.editor.document.selection.getSelectedBlocks() );
+
+		if ( !block ) {
+			return false;
 		}
+
+		const schema = this.editor.document.schema;
+		const isParagraphAllowed = schema.check( {
+			name: 'paragraph',
+			inside: Position.createBefore( block )
+		} );
+
+		return isParagraphAllowed && !schema.objects.has( block.name );
```

Comment:
I know we talked about this, but I'm not sure about the last bit. Why checking whether the block is an object too? That doesn't say yet that it cannot be turned into a paragraph. Or rather – it's an indication, just like being in `schema.limits` (which is even clearer indication). However, in the second case, the element won't be returned by `getSelectedBlocks()` because only blocks are returned by this method. So I think now that this is unnecessary here.

Besides, if you make this change here, we'd need to make it in all other similar commands. So I'd rather like to have it in one place – `getSelectedBlocks()`.